examples/erc1155/lib.rs

```diff
@@ -0,0 +1,687 @@
+    /// Any callers must revert if they receive anything other than `BATCH_MAGIC_VALUE` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_batch_received(
+        &mut self,
+        operator: AccountId,
+        from: AccountId,
+        token_ids: Vec<TokenId>,
+        values: Vec<Balance>,
+        data: Vec<u8>,
+    );
+}
+
+#[ink::contract]
+mod erc1155 {
+    use super::*;
+
+    use ink_env::call::{
+        build_call,
+        utils::ReturnType,
+        ExecutionInput,
+        Selector,
+    };
+    use ink_prelude::collections::BTreeMap;
+
+    /// Indicate that a token transfer has occured.
+    ///
+    /// This must be emitted even if a zero value transfer occurs.
+    #[ink(event)]
+    pub struct TransferSingle {
+        #[ink(topic)]
+        operator: AccountId,
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+        token_id: TokenId,
+        value: Balance,
+    }
+
+    /// Indicate that an approval event has happened.
+    #[ink(event)]
+    pub struct ApprovalForAll {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        operator: AccountId,
+        approved: bool,
+    }
+
+    /// Indicate that a token's URI has been updated.
+    #[ink(event)]
+    pub struct Uri {
+        value: ink_prelude::string::String,
+        #[ink(topic)]
+        token_id: TokenId,
+    }
+
+    /// An ERC-1155 contract.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Contract {
+        /// Tracks the balances of accounts across the different tokens that they might be holding.
+        balances: BTreeMap<(AccountId, TokenId), Balance>,
+        /// Which accounts (called operators) have been approved to spend funds on behalf of an owner.
+        ///
+        /// Note that the mapping is Set<(Owner, Operator)>.
+        approvals: BTreeMap<(AccountId, AccountId), ()>,
+        /// A unique identifier for the tokens which have been minted (and are therefore supported)
+        /// by this contract.
+        token_id_nonce: TokenId,
+    }
+
+    impl Contract {
+        /// Initialize a default instance of this ERC-1155 implementation.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Default::default()
+        }
+
+        /// Create the initial supply for a token.
+        ///
+        /// The initial supply will be provided to the caller (a.k.a the minter), and the
+        /// `token_id` will be assigned by the smart contract.
+        ///
+        /// Note that as implemented anyone can create tokens. If you were to deploy this to a
+        /// production environment you'd probably want to lock down the addresses that are allowed
+        /// to create tokens.
+        #[ink(message)]
+        pub fn create(&mut self, value: Balance, _data: Vec<u8>) -> TokenId {
+            // Given that TokenId is a `u128` the likelihood of this overflowing is pretty slim.
+            self.token_id_nonce += 1;
```

Comment:
I see a problem with respect to contract economics even in the case where we restrict the set of people who are able to create new tokens because it seems that there is no way to delete coins altogether and therefore free their occupied space.

Reply:
We have the `ink_storage::Stash` data structure for these kind of purposes. However, it is a bit more complex that you need here. We might want to add yet another bit-wise data structure for use cases like these.

Reply:
> there is no way to delete coins altogether and therefore free their occupied space.

I don't know if this is a good "feature" to have tbh. As a token holder I (probably) wouldn't want someone/something to be able to clear my balance in order to save some space on-chain.

Maybe a use case here is that project is transitioning to a new token, in which case you'd want to burn the old token and eventually clear those balance entries. However, this whole mechanism seems very implementation specific